### PR TITLE
feat: slow php reporting page

### PR DESCRIPTION
**Pull request by KrawczowaKris**
# Summary
## Test Task
**Slow php reporting page**

A typical scenario:
-  you have a web app
-  that app has a backend written in PHP
-  most of the backend works fast and good for end users < 1000ms
-  however there is a slow page /reports/slow, that takes 10m to execute, which causes nginx and php to timeout
-  implement such solution so that nginx/php-fpm do not timeout and get the report when 10m passed.

Implementation nuances:
-  your nginx and php-fpm should run under nginx/nginx user/group.
-  use sleep php func to simulate such report, https://www.php.net/manual/en/function.sleep.php
-  implement a custom location /reports/fast, that will work fast and doesn't require any timeouts, and will use www pool of the PHP-FPM
-  implement a custom location /reports/slow, that will open /tmp/nginx/reports/slow.php script, which will take 10m to execute
-  during the /reports/slow, you should not lose the browser connection to the server.
-  IMPORTANT: you can't change any values for your core application that runs inside www pool defined in the PHP-FPM, for example you can't change timeout via www pool, as that may substantially degrade all the other URLs that are working good and fast - so think how you can do it.

When you open PR - pls package your solution as zip file with bash script run.sh that should
-  create all necessary files and/or folders.
-  set proper permissions to created files/folders
-  start nginx in foreground mode (so all errors, logs are printed on stdout/stderr)
-  start php-fpm in foreground mode (so all errors, logs are printed on stdout/stderr)
-  curl https://localhost/reports/fast should output "FAST REPORT" via PHP script
-  curl https://localhost/reports/slow should output "SLOW REPORT" via PHP script with 10m delay without losing any connection (i.e. no timeout or upstream timeout error in the nginx)

## my results
[php-slow.zip](https://github.com/KrawczowaKris/devops-camp-2023/files/11293585/php-slow.zip)

./run.sh
**output:**
```
FAST REPORT
real	0m0,069s
user	0m0,064s
sys	0m0,000s
SLOW REPORT
real	10m0,107s
user	0m0,071s
sys	0m0,027s
```

### CODE REVIEW

File: chapter3/nginx.conf [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+			fastcgi_param  REMOTE_ADDR        $remote_addr;
+			fastcgi_param  REMOTE_PORT        $remote_port;
+			fastcgi_param  SERVER_ADDR        $server_addr;
+			fastcgi_param  SERVER_PORT        $server_port;
+			fastcgi_param  SERVER_NAME        $server_name;
+			fastcgi_param  REDIRECT_STATUS    200;
+			fastcgi_param  PATH_INFO          $fastcgi_path_info;
+			
+			fastcgi_pass   reports-fast;
+			fastcgi_param SCRIPT_FILENAME /var/www/reports/fast.php;
```
Comment (dmitry-mightydevops) [inline chapter3/nginx.conf:61]: don't use hardcoded value. Find a different approach and use a realpath I have mentioned in the lecture.

File: chapter3/nginx.conf [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+			fastcgi_param  REMOTE_ADDR        $remote_addr;
+			fastcgi_param  REMOTE_PORT        $remote_port;
+			fastcgi_param  SERVER_ADDR        $server_addr;
+			fastcgi_param  SERVER_PORT        $server_port;
+			fastcgi_param  SERVER_NAME        $server_name;
+			fastcgi_param  REDIRECT_STATUS    200;
+			fastcgi_param  PATH_INFO          $fastcgi_path_info;
+			
+			fastcgi_pass   reports-slow;
+			fastcgi_param SCRIPT_FILENAME /var/www/reports/slow.php;
```
Comment (dmitry-mightydevops) [inline chapter3/nginx.conf:91]: Same here

File: chapter3/php-slow.zip
Comment (dmitry-mightydevops) [inline chapter3/php-slow.zip:1]: all binaries should not be pushed in the remote. Attach them to git PR summary.

File: chapter3/php-fpm.conf [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+[www]
+user = nginx
+group = nginx
+listen = /run/php/www.sock
+listen.owner = nginx
+listen.group = nginx
+pm = static
+pm.max_children = 1
+
+[slow]
```
Comment (dmitry-mightydevops) [inline chapter3/php-fpm.conf:16]: that is not a correct solution as your "slow" pool has no difference than the www pool. Your timeout group of settings have to be defined as `admin` type value in php-fpm.conf and as well as in the location.

File: chapter3/php.ini [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+doc_root =
+user_dir =
+enable_dl = Off
+file_uploads = On
+upload_max_filesize = 2M
+max_file_uploads = 20
+allow_url_fopen = On
+allow_url_include = Off
+default_socket_timeout = 60
+max_execution_timeout = 630s
```
Comment (dmitry-mightydevops) [inline chapter3/php.ini:47]: that has to exists in the php-fpm.conf in a proper pool. Right now you allow ALL SCRIPTS to be super slow.
In our case we still need to timeout in 30s (or 60s) except for the slow pool (that we know ahead of time is always slow)

File: chapter3/run.sh [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+sudo chown -R nginx:nginx /tmp/localhost.pem
+sudo chown -R nginx:nginx /tmp/localhost-key.pem
+sudo chmod 777 /var/log
+
+# Launch nginx and php-fpm
+sudo fuser -k 443/tcp
+sudo /usr/sbin/php-fpm8.2 --php-ini "$(pwd)/php.ini" -y "$(pwd)/php-fpm.conf"
+sudo nginx -c "$(pwd)/nginx.conf"
+
+sleep 3
```
Comment (dmitry-mightydevops) [inline chapter3/run.sh:18]: no need for sleep. Check with `while cycle, using CURL call` that nginx is up & running.

File: chapter3/run.sh [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+
+# Sending requests
+curl https://localhost/reports/fast
```
Comment (dmitry-mightydevops) [inline chapter3/run.sh:21]: learn how to use `time`, i.e. `time curl <URL>`

File: chapter3/php.ini (hunk 4 above)
Comment (dmitry-mightydevops) [reply chapter3/php.ini:47]: So keep in mind that php.ini is a global file and all settings from php.ini will be applied to all PHP processes until you override them in php-fpm.conf via `value` flags.

**Review by dmitry-mightydevops — CHANGES_REQUESTED** [review]

File: chapter3/php.ini [hunk 7]
```diff
@@ -0,0 +1,169 @@
...
+implicit_flush = Off
+unserialize_callback_func =
+serialize_precision = -1
+disable_functions =
+disable_classes =
+zend.enable_gc = On
+zend.exception_ignore_args = On
+zend.exception_string_param_max_len = 0
+expose_php = On
+max_execution_time = 30
```
Comment (dmitry-mightydevops) [inline chapter3/php.ini:19]: you duplicate the values. That will have side effects

**Review by dmitry-mightydevops — CHANGES_REQUESTED** [review]

File: chapter3/php-slow.zip
Comment (KrawczowaKris) [reply chapter3/php-slow.zip:1]: done

File: chapter3/php.ini (hunk 4 above)
Comment (KrawczowaKris) [reply chapter3/php.ini:47]: remade

File: chapter3/php.ini (hunk 7 above)
Comment (KrawczowaKris) [reply chapter3/php.ini:19]: remade

File: chapter3/nginx.conf (hunk 1 above)
Comment (KrawczowaKris) [reply chapter3/nginx.conf:61]: done

File: chapter3/nginx.conf (hunk 2 above)
Comment (KrawczowaKris) [reply chapter3/nginx.conf:91]: done

File: chapter3/php-fpm.conf (hunk 3 above)
Comment (KrawczowaKris) [reply chapter3/php-fpm.conf:16]: done

File: chapter3/run.sh (hunk 6 above)
Comment (KrawczowaKris) [reply chapter3/run.sh:21]: done

File: chapter3/nginx.conf [hunk 8]
```diff
@@ -87,8 +85,8 @@ http {
 			fastcgi_param  REDIRECT_STATUS    200;
 			fastcgi_param  PATH_INFO          $fastcgi_path_info;
 			
-			fastcgi_pass   reports-slow;
-			fastcgi_param SCRIPT_FILENAME /var/www/reports/slow.php;
+			fastcgi_pass reports;
```
Comment (dmitry-mightydevops) [inline chapter3/nginx.conf:88]: your slow page should `fastcg_pass` into slow pool. You should have two different pools per 2 locations

File: chapter3/php-fpm.conf [hunk 9]
```diff
@@ -12,12 +12,3 @@ listen.owner = nginx
...
 pm.max_children = 1
-
-[slow]
-user = nginx
-group = nginx
-listen = /run/php/slow.sock
-listen.owner = nginx
-listen.group = nginx
-pm = static
-pm.max_children = 1
```
Comment (dmitry-mightydevops) [inline chapter3/php-fpm.conf]: why removed?

**Review by dmitry-mightydevops — CHANGES_REQUESTED** [review]
you misunderstood my previous comments - pls re-read them. Right now implementation is not correct. Your previous implementation was right path. This one is not.

File: chapter3/php-fpm.conf (hunk 3 above)
Comment (dmitry-mightydevops) [reply chapter3/php-fpm.conf]: @KrawczowaKris you should not remove the slow pool. Slow pool should define proper timeouts for workers using `admin` calls to override changes in php.ini.

**Review by dmitry-mightydevops — CHANGES_REQUESTED** [review]
and you still have duplicates in php.ini

php.ini should not have duplicated instructions even if the value is the same
image

that lead to side effects. Someone will modify the top value and will not notice the bottom value exists - as a result it will be still 30s.

File: chapter3/run.sh [hunk 10]
```diff
@@ -0,0 +1,21 @@
...
+sudo chown -R nginx:nginx /tmp/localhost-key.pem
+sudo chmod 777 /var/log
+
+# Launch nginx and php-fpm
+sudo fuser -k 443/tcp
+sudo /usr/sbin/php-fpm8.2 --php-ini "$(pwd)/php.ini" -y "$(pwd)/php-fpm.conf"
+sudo nginx -c "$(pwd)/nginx.conf"
+
+# Sending requests
+curl -w " %{time_total}" https://localhost/reports/fast
```
Comment (dmitry-mightydevops) [inline chapter3/run.sh:19]: use time

File: chapter3/run.sh [hunk 11]
```diff
@@ -0,0 +1,21 @@
...
+echo
+curl -w " %{time_total}" https://localhost/reports/slow
```
Comment (dmitry-mightydevops) [inline chapter3/run.sh:21]: use time

Comment (dmitry-mightydevops) [reply chapter3/run.sh:21]:
```
on ☸ crossingguard-staging-eks (ci) saritasa-devops-camp on 🌱 main [?] 
➜ time curl -I 'Host: theproducebox.com'  https://54.235.197.83 --insecure
curl: (3) URL using bad/illegal format or missing URL
HTTP/1.1 301 Moved Permanently
Date: Mon, 24 Apr 2023 20:29:57 GMT
Server: Apache
X-Powered-By: PHP/7.4.15
Expires: Thu, 19 Nov 1981 08:52:00 GMT
Cache-Control: no-store, no-cache, must-revalidate
Pragma: no-cache
X-Pingback: https://theproducebox.com/xmlrpc.php
X-Redirect-By: WordPress
Set-Cookie: PHPSESSID=brbhv92f99030s2ujahv8lpqet; path=/
X-Frame-Options: SAMEORIGIN
Location: https://theproducebox.com/
Content-Type: text/html; charset=UTF-8

curl -I 'Host: theproducebox.com' https://54.235.197.83 --insecure  0.01s user 0.01s system 2% cpu 0.558 total

```

in this case we get `0.01s user 0.01s system 2% cpu 0.558 total` - which is what we need

**Review by dmitry-mightydevops — CHANGES_REQUESTED** [review]

File: chapter3/run.sh (hunk 6 above)
Comment (KrawczowaKris) [reply chapter3/run.sh]: done

File: chapter3/php-fpm.conf (hunk 3 above)
Comment (KrawczowaKris) [reply chapter3/php-fpm.conf]: remade

File: chapter3/nginx.conf (hunk 8 above)
Comment (KrawczowaKris) [reply chapter3/nginx.conf:88]: remade

File: chapter3/php-fpm.conf (hunk 9 above)
Comment (KrawczowaKris) [reply chapter3/php-fpm.conf]: remade

File: chapter3/run.sh (hunk 10 above)
Comment (KrawczowaKris) [reply chapter3/run.sh:19]: remade

File: chapter3/run.sh (hunk 11 above)
Comment (KrawczowaKris) [reply chapter3/run.sh:21]: remade

File: chapter3/run.sh (hunk 5 above)
Comment (KrawczowaKris) [reply chapter3/run.sh]: did not quite understand what is needed

File: chapter3/php-fpm.conf [hunk 12]
```diff
@@ -12,3 +12,14 @@ listen.owner = nginx
...
+
+[slow]
+user = nginx
+group = nginx
+listen = /run/php/slow.sock
+listen.owner = nginx
+listen.group = nginx
+pm = static
+pm.max_children = 1
+request_terminate_timeout=630s
```
Comment (dmitry-mightydevops) [inline chapter3/php-fpm.conf:25]: that is a different setting, which will KILL the worker after the desired timeout - we don't want the kill. Look into php.ini max_execution_time and set it via `php_admin_value`
https://code.tutsplus.com/tutorials/how-to-increase-max_execution_time-in-php--cms-37017

Comment (dmitry-mightydevops) [reply chapter3/php-fpm.conf:25]: 1. so you set max_execution_time in fast pool = 60s  and you set max_execution_time = 600s in the slow pool
2. then you need to set [fastcgi_read_timeout](https://nginx.org/en/docs/http/ngx_http_fastcgi_module.html#fastcgi_read_timeout) to 60 and 600 in the respective server/location contexts.

**Review by dmitry-mightydevops — CHANGES_REQUESTED** [review]

File: chapter3/run.sh (hunk 5 above)
Comment (dmitry-mightydevops) [reply chapter3/run.sh]: so your run.sh will not work if for example nginx started in 5s. That's why you need to have a `while` block in the code.

Comment (KrawczowaKris) [reply chapter3/run.sh]: added

File: chapter3/php-fpm.conf (hunk 12 above)
Comment (KrawczowaKris) [reply chapter3/php-fpm.conf]: remade

File: chapter3/php.ini [hunk 13]
```diff
@@ -44,7 +40,7 @@ max_file_uploads = 20
 allow_url_fopen = On
 allow_url_include = Off
 default_socket_timeout = 60
-max_execution_timeout = 630s 
+php_admin_value max_execution_timeout = 60s
```
Comment (dmitry-mightydevops) [inline chapter3/php.ini:43]: no need to have this in php.ini
only in php-fpm pool

**Review by dmitry-mightydevops — CHANGES_REQUESTED** [review]

File: chapter3/php.ini (hunk 13 above)
Comment (KrawczowaKris) [reply chapter3/php.ini]: done

**Review by dmitry-mightydevops — APPROVED** [review]

**Pull request merged**